<a href="https://colab.research.google.com/github/zarfer007/AprendizajePorRefuerzos/blob/master/testingGit.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# RL 2 LAB I.1

  - Morbidoni, Emilio
  - Zarzosa, Fernando


# Interfaz básica stable-baselines

### Instalación de Stable-baselines

  - Desde Windows, además, instalar: 
* Microsoft Visual C++ desde https://visualstudio.microsoft.com/visual-cpp-build-tools/
* PyType, mediante `conda install -c conda-forge pytype`

  - Desde Linux o Google Colab

In [ ]:
#!pip install sphinxcontrib-spelling # if problems with sphinx
#!pip install datascience

In [ ]:
#@title Instalación (no modificar)
!pip install stable-baselines3[extra,tests,docs]>=0.11.0a4 && pip install sb3-contrib

ERROR: datascience 0.10.6 has requirement coverage==3.7.1, but you'll have coverage 5.5 which is incompatible.
ERROR: datascience 0.10.6 has requirement folium==0.2.1, but you'll have folium 0.8.3 which is incompatible.
ERROR: coveralls 0.5 has requirement coverage<3.999,>=3.6, but you'll have coverage 5.5 which is incompatible.
ERROR: sphinxcontrib-spelling 7.1.0 has requirement Sphinx>=3.0.0, but you'll have sphinx 1.8.5 which is incompatible.
ERROR: sphinx-autodoc-typehints 1.11.1 has requirement Sphinx>=3.0, but you'll have sphinx 1.8.5 which is incompatible.
ERROR: pytest-cov 2.11.1 has requirement pytest>=4.6, but you'll have pytest 3.6.4 which is incompatible.
ERROR: pytest-forked 1.3.0 has requirement pytest>=3.10, but you'll have pytest 3.6.4 which is incompatible.
ERROR: pytest-xdist 2.2.1 has requirement pytest>=6.0.0, but you'll have pytest 3.6.4 which is incompatible.
  Using cached https://files.pythonhosted.org/packages/f6/62/796d8ae02732c162f8d53406f520c9f3c886a9ab24de4

# RL-baselines3-zoo

  - Instalación de RLBaselinesZoo

Desde Google Colab

  - ---recursive allows to train agents

In [ ]:
#@title Instalación de RLBaselinesZoo (no modificar)

#if IN_COLAB:
!git clone --recursive https://github.com/DLR-RM/rl-baselines3-zoo
!cd rl-baselines3-zoo/
!apt-get install swig cmake ffmpeg
!pip install -r /content/rl-baselines3-zoo/requirements.txt

Cloning into 'rl-baselines3-zoo'...
remote: Enumerating objects: 275, done.
remote: Counting objects: 100% (275/275), done.
remote: Compressing objects: 100% (143/143), done.
remote: Total 2161 (delta 18), reused 257 (delta 12), pack-reused 1886
Receiving objects: 100% (2161/2161), 1.00 MiB | 14.09 MiB/s, done.
Resolving deltas: 100% (1363/1363), done.
Submodule 'rl-trained-agents' (https://github.com/DLR-RM/rl-trained-agents) registered for path 'rl-trained-agents'
Cloning into '/content/rl-baselines3-zoo/rl-trained-agents'...
remote: Enumerating objects: 1, done.        
remote: Counting objects: 100% (1/1), done.        
remote: Total 1416 (delta 0), reused 0 (delta 0), pack-reused 1415
Receiving objects: 100% (1416/1416), 977.06 MiB | 32.34 MiB/s, done.
Resolving deltas: 100% (220/220), done.
Submodule path 'rl-trained-agents': checked out 'd81fcd61cef4599564c859297ea68bacf677db6b'
Reading package lists... Done
Building dependency tree       
Reading state information... Done
cmake

Desde Linux, ejecutando

    git clone --recursive https://github.com/DLR-RM/rl-baselines3-zoo
    cd rl-baselines3-zoo/
    sudo apt-get install swig cmake ffmpeg
    pip install -r requirements.txt

Instalación desde Nabucodonosor

    wget https://repo.anaconda.com/archive/Anaconda3-2020.11-Linux-x86_64.sh
    chmod 755 Anaconda3-2020.11-Linux-x86_64.sh
    ./Anaconda3-2020.11-Linux-x86_64.sh
    conda create --name rl
    conda activate rl
    conda config --add channels conda-forge
    conda install jupyter atari_py swig
    pip install stable-baselines3[extra,tests,docs]>=0.11.0a4
    pip install sb3-contrib 

## Ejecución de un algoritmo de RL
  - Importaciones/inicializaciones

In [ ]:
import os
from subprocess import Popen, PIPE

import numpy as np
import matplotlib.pyplot as plt

import gym
from gym import spaces
#from gym.envs.registration import register

from stable_baselines3 import DQN, PPO
from stable_baselines3.common.monitor import Monitor
from stable_baselines3.common.vec_env import DummyVecEnv, VecNormalize
from stable_baselines3.common.callbacks import EvalCallback, StopTrainingOnRewardThreshold
from stable_baselines3.common.env_util import make_vec_env

os.makedirs('logs', exist_ok=True)

try:
  import google.colab
  IN_COLAB = True
except:
  IN_COLAB = False

cwd = os.getcwd()

%matplotlib inline
%load_ext tensorboard

Para usar un entorno compatible por esta librería, el mismo tiene que heredar de *gym.Env*. Vemos un ejemplo (crédito: https://colab.research.google.com/github/araffin/rl-tutorial-jnrr19/blob/sb3/5_custom_gym_env.ipynb)

# Lab 2

1. Crear tu propio entorno y entrenar agentes RL en el mismo. Analizar la convergencia con distintos algoritmos* (ej: PPO, DQN), resultados con distintas funciones de recompensa e híper-parámetros. 

    Algunas ideas:

    * Transformar GoLeftEnv en una grilla 2D, añadir paredes / trampas / agua.
    * Crear un entorno que juegue a algún juego como el ta-te-ti.
    * Crea un entorno totalmente nuevo que sea de tu interés!

2. Entrena agentes en entornos más complejos con stable-baselines/rl-baselines-zoo. Tener en cuenta:

    * Google Colab tiene una limitante en cuanto a cantidad de recursos de CPU/GPU (incluido un "rendimiento decreciente silencioso"), lo cuál reduce la capacidad de entrenar distintos entornos.
    * Si el entorno no está implementado en stable-baselines, debe hacerse un wrapper a mano, lo que puede ser sencillo o puede llevar algo más de trabajo, teniendo que tocar código subyacente de la librería. 

\* pueden ser usando stable-baselines/rl-baselines-zoo o bien utilizando algún otro algoritmo (incluso tabular)

# 2.2 Training

## Ejecución

Los agentes pueden ser llamados desde la consola mediante comandos como

`python train.py --algo algo_name --env env_id`

Los cuales pueden ser llamados usando


También es posible grabar un video! Ver https://stable-baselines3.readthedocs.io/en/master/guide/examples.html#record-a-video

# MountainCarContinuous

In [ ]:
!apt-get update
!apt-get install swig cmake libopenmpi-dev zlib1g-dev ffmpeg freeglut3-dev xvfb
!pip install stable-baselines[mpi] --upgrade
!pip install pybullet
!pip install box2d box2d-kengz pyyaml pytablewriter optuna scikit-optimize
# Stable Baselines only supports tensorflow 1.x for now
%tensorflow_version 1.x

Hit:1 http://security.ubuntu.com/ubuntu bionic-security InRelease
Ign:2 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease
Hit:3 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ InRelease
Hit:4 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic InRelease
Hit:5 http://archive.ubuntu.com/ubuntu bionic InRelease
Ign:6 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  InRelease
Hit:7 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Release
Hit:8 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  Release
Hit:9 http://archive.ubuntu.com/ubuntu bionic-updates InRelease
Hit:10 http://ppa.launchpad.net/cran/libgit2/ubuntu bionic InRelease
Hit:11 http://archive.ubuntu.com/ubuntu bionic-backports InRelease
Hit:12 http://ppa.launchpad.net/deadsnakes/ppa/ubuntu bionic InRelease
Hit:13 http://ppa.launchpad.net/graphics-drivers/ppa/ubuntu bionic 

In [ ]:
!git clone https://github.com/araffin/rl-baselines-zoo


Cloning into 'rl-baselines-zoo'...
remote: Enumerating objects: 3, done.
remote: Counting objects: 100% (3/3), done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 1843 (delta 0), reused 1 (delta 0), pack-reused 1840
Receiving objects: 100% (1843/1843), 375.67 MiB | 33.32 MiB/s, done.
Resolving deltas: 100% (1091/1091), done.
Checking out files: 100% (333/333), done.


In [ ]:
cd rl-baselines-zoo/

/content/rl-baselines-zoo/rl-baselines-zoo


### Source:
https://gym.openai.com/envs/MountainCarContinuous-v0/

In [ ]:
tensorflow_version # tensorflow version for gym is 1.x

Unknown TensorFlow version: # tensorflow version for gym is 1.x
Currently selected TF version: 1.x
Available versions:
 * 1.x
 * 2.x


In [ ]:
# Baseline Model
!python train.py --algo ppo2 --env MountainCarContinuous-v0

The TensorFlow contrib module will not be included in TensorFlow 2.0.
For more information, please see:
  * https://github.com/tensorflow/community/blob/master/rfcs/20180907-contrib-sunset.md
  * https://github.com/tensorflow/addons
  * https://github.com/tensorflow/io (for I/O related ops)
If you depend on functionality not listed there, please file an issue.


========== MountainCarContinuous-v0 ==========
Seed: 0
OrderedDict([('ent_coef', 0.0),
             ('gamma', 0.99),
             ('lam', 0.94),
             ('n_envs', 16),
             ('n_steps', 256),
             ('n_timesteps', 1000000.0),
             ('nminibatches', 8),
             ('noptepochs', 4),
             ('normalize', True),
             ('policy', 'MlpPolicy')])
Using 16 environments
Normalizing input and reward
Creating test environment
Normalization activated: {'norm_reward': False}




Instructions for updating:
Use keras.layers.flatten instead.
Instructions for updating:
Please use `layer.__call__` metho

In [ ]:
# Overwritting the baseline timesteps model

!python train.py --algo ppo2 --env MountainCarContinuous-v0 --n-timesteps 10000 #--save-freq 1000


The TensorFlow contrib module will not be included in TensorFlow 2.0.
For more information, please see:
  * https://github.com/tensorflow/community/blob/master/rfcs/20180907-contrib-sunset.md
  * https://github.com/tensorflow/addons
  * https://github.com/tensorflow/io (for I/O related ops)
If you depend on functionality not listed there, please file an issue.


========== MountainCarContinuous-v0 ==========
Seed: 0
OrderedDict([('ent_coef', 0.0),
             ('gamma', 0.99),
             ('lam', 0.94),
             ('n_envs', 16),
             ('n_steps', 256),
             ('n_timesteps', 1000000.0),
             ('nminibatches', 8),
             ('noptepochs', 4),
             ('normalize', True),
             ('policy', 'MlpPolicy')])
Using 16 environments
Overwriting n_timesteps with n=10000
Normalizing input and reward
Creating test environment
Normalization activated: {'norm_reward': False}




Instructions for updating:
Use keras.layers.flatten instead.
Instructions for updati

## Optimizing search of parameters with Optuna

  - link of optuna: Used for changing hyperparameter (looking for the best so it optimizes a search)
  - https://colab.research.google.com/github/optuna/optuna/blob/master/examples/quickstart.ipynb
  - Research related to the algorithm Optuna (https://arxiv.org/pdf/1907.10902.pdf)
  - That is why we add an hyperparameter optimizer search (up to the search of 10 parameters based on a n (=300, in our exercise) sample trial
  - There are many selection of hyperparameters that can be chosen, but in terms of cost-effectiveness some could be skipped (or prunned).

In [ ]:
#%tensorboard --logdir=tensorboard
!python train.py --algo ppo2 --env MountainCarContinuous-v0 -n 50000 -optimize --n-trials 300 --n-jobs 2 --sampler tpe --pruner median


The TensorFlow contrib module will not be included in TensorFlow 2.0.
For more information, please see:
  * https://github.com/tensorflow/community/blob/master/rfcs/20180907-contrib-sunset.md
  * https://github.com/tensorflow/addons
  * https://github.com/tensorflow/io (for I/O related ops)
If you depend on functionality not listed there, please file an issue.


========== MountainCarContinuous-v0 ==========
Seed: 0
OrderedDict([('ent_coef', 0.0),
             ('gamma', 0.99),
             ('lam', 0.94),
             ('n_envs', 16),
             ('n_steps', 256),
             ('n_timesteps', 1000000.0),
             ('nminibatches', 8),
             ('noptepochs', 4),
             ('normalize', True),
             ('policy', 'MlpPolicy')])
Using 16 environments
Overwriting n_timesteps with n=50000
Normalizing input and reward
Optimizing hyperparameters
Sampler: tpe - Pruner: median
[I 2021-03-19 14:44:52,481] A new study created in memory with name: no-name-35872e91-0846-4263-8b8e-ea34

### Optimal combination of hyperparameters for 300 trials
  - We run a search model with 1000 trials, up to the 376 model the algorithm stop due to runtime out. It suggested then that the best combination was given by the trial 115
  - To avoid runtime probles we ran the model with 300 trials. The best model chosen was the one of 115 trials (see description below)
  - We cannot see the model running, but we asume the code (!python train.py --algo ppo2 --env MountainCarContinuous-v0 -n 50000 -optimize --n-trials 300 --n-jobs 2 --sampler tpe --pruner median) does not only search the optimal combination of hyperparameters, but it also trains the adapted model (as the !python suggests)
  - The only addition to the !python... order that worked, including the optimised set of hyperparameters) was: !python train.py --algo ppo2 --env MountainCarContinuous-v0 --n-timesteps 256 #--n_envs 16 

In [ ]:

'''
Optuna search results, 300 trials
Params: 
    batch_size: 32
    n_steps: 256
    gamma: 0.95
    lr: 0.0006213010340576837
    ent_coef: 0.0004927687872681538
    cliprange: 0.2
    noptepochs: 5
    lambda: 1.0
'''
!python train.py --algo ppo2 --env MountainCarContinuous-v0 --n-timesteps 256 #--n_envs 16

The TensorFlow contrib module will not be included in TensorFlow 2.0.
For more information, please see:
  * https://github.com/tensorflow/community/blob/master/rfcs/20180907-contrib-sunset.md
  * https://github.com/tensorflow/addons
  * https://github.com/tensorflow/io (for I/O related ops)
If you depend on functionality not listed there, please file an issue.


========== MountainCarContinuous-v0 ==========
Seed: 0
OrderedDict([('ent_coef', 0.0),
             ('gamma', 0.99),
             ('lam', 0.94),
             ('n_envs', 16),
             ('n_steps', 256),
             ('n_timesteps', 1000000.0),
             ('nminibatches', 8),
             ('noptepochs', 4),
             ('normalize', True),
             ('policy', 'MlpPolicy')])
Using 16 environments
Overwriting n_timesteps with n=256
Normalizing input and reward
Creating test environment
Normalization activated: {'norm_reward': False}




Instructions for updating:
Use keras.layers.flatten instead.
Instructions for updating

In [ ]:
'''
# Tuned
MountainCarContinuous-v0:
  normalize: true
  n_envs: 1
  n_timesteps: !!float 20000
  policy: 'MlpPolicy'
  batch_size: 256
  n_steps: 8
  gamma: 0.9999
  learning_rate: !!float 7.77e-05
  ent_coef: 0.00429
  clip_range: 0.1
  n_epochs: 10
  gae_lambda: 0.9
  max_grad_norm: 5
  vf_coef: 0.19
  use_sde: True
  policy_kwargs: "dict(log_std_init=-3.29, ortho_init=False)"
'''
#!python train.py --algo trpo --env HotterColder-v0 # it is not in training format yet
# algorithms a2c,acer,acktr,dqn,ddpg,her,sac,ppo2,trpo,td3